WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift

```diff
@@ -0,0 +1,111 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireAPI
+import WireDataModel
+
+// sourcery: AutoMockable
+/// Facilitate access to conversations related domain objects.
+public protocol ConversationRepositoryProtocol {
+
+    /// Fetches and persists all conversations
+
+    func pullConversations() async throws
+
+}
+
+public final class ConversationRepository: ConversationRepositoryProtocol {
+
+    public struct BackendInfo {
+        let domain: String
+        let isFederationEnabled: Bool
+    }
+
+    // MARK: - Properties
+
+    private let conversationsAPI: any ConversationsAPI
+    private let conversationsLocalStore: any ConversationLocalStoreProtocol
+    private let backendInfo: BackendInfo
+
+    // MARK: - Object lifecycle
+
+    public init(
+        conversationsAPI: any ConversationsAPI,
+        conversationsLocalStore: any ConversationLocalStoreProtocol,
+        backendInfo: BackendInfo
+    ) {
+        self.conversationsAPI = conversationsAPI
+        self.conversationsLocalStore = conversationsLocalStore
+        self.backendInfo = backendInfo
+    }
+
+    // MARK: - Public
+
+    public func pullConversations() async throws {
+        var qualifiedIds: [WireAPI.QualifiedID]
+
+        if let result = try? await conversationsAPI.getLegacyConversationIdentifiers() { /// only for api v0 (see `ConversationsAPIV0` method comment)
+            let uuids = try await result.reduce([UUID](), +)
+            qualifiedIds = uuids.map { WireAPI.QualifiedID(uuid: $0, domain: backendInfo.domain) }
+        } else {
+            /// fallback to api versions > v0.
+            let ids = try await conversationsAPI.getConversationIdentifiers()
+            qualifiedIds = try await ids.reduce([WireAPI.QualifiedID](), +)
+        }
+
+        let conversationList = try await conversationsAPI.getConversations(for: qualifiedIds)
+
+        await withThrowingTaskGroup(of: Void.self) { taskGroup in
+            let foundConversations = conversationList.found
+            let missingConversationsQualifiedIds = conversationList.notFound
+            let failedConversationsQualifiedIds = conversationList.failed
+
+            for conversation in foundConversations {
+                taskGroup.addTask { [self] in
+                    do {
+                        try await conversationsLocalStore.storeConversation(
+                            conversation,
+                            isFederationEnabled: backendInfo.isFederationEnabled
+                        )
+                    } catch {
+                        throw ConversationRepositoryError.failedToStoreConversation(error)
```

Comment:
not sure here, does it break the loop at first error?

Reply:
Interestingly enough, it would do nothing because we are throwing inside a child task and not inside the body of `withThrowingTaskGroup` itself. If we were throwing inside the body, it would cancel the group and all of its child tasks but here it will do nothing and continue processing. I'm not a 100% sure but looking at Apple's doc and some online resources, it looks like it. Bottom line, if we want to `actually` throw and propagate the error to the call site, it should be done inside the body of `withThrowingTaskGroup` and not inside the child task.

Anyways, this was removed now that I added back the logger, we are catching the underlying error and logging it, so we don't need to have a throwing func anymore.